File: Gombang/[LeetCode75] - [DFS]/437_Path_Sum_III.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    int pathCount = 0;

    void dfs(TreeNode* root, vector<long long> vec, int targetSum)
    {
        if (root == nullptr)
            return;

        for (long long& value : vec)
        {
            value += root->val;
        }

        vec.push_back(root->val);

        for (const long long& value : vec)
        {
            if (value == targetSum)
                pathCount++;
        }

        dfs(root->left, vec, targetSum);
        dfs(root->right, vec, targetSum);
    }

    int pathSum(TreeNode* root, int targetSum) {
        // 현재 노드까지의 합과 targetSum을 비교할 수 있는 알고리즘 짜기.

        vector<long long> vec;
        dfs(root, vec, targetSum);

        return pathCount;
    }
};

File: Seori/LeetCode 75/Binary Tree_DFS/437. Path Sum III.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        def dfs(root: Optional[TreeNode], level: int, listSum: List) -> int:
            # 재귀 종료조건
            if root == None:
                return 0
            
            # 매 노드를 탐색하는 시점에 구간 누적합을 체크
            count = 0
            value = root.val
            new_listSum = listSum[:]
            new_listSum.append(0)
            for i in range(level):
                new_listSum[i] += value
                if new_listSum[i] == targetSum:
                    count += 1
            
            # 전위순회
            count += dfs(root.left, level + 1, new_listSum)
            count += dfs(root.right, level + 1, new_listSum)

            return count

        return dfs(root, 1, [])